table_engine/src/provider.rs

```diff
@@ -122,12 +124,61 @@ impl TableProviderAdapter {
         Ok(Arc::new(scan_table))
     }
 
-    fn predicate_from_filters(&self, filters: &[Expr]) -> PredicateRef {
+    fn check_and_build_predicate_from_filters(&self, filters: &[Expr]) -> PredicateRef {
+        // 1.Only filters of primary key or timestamp key can be pushed down.
+        // Build key set.
+        let mut key_set = HashSet::new();
+        // Timestamp key should be always picked.
+        key_set.insert(self.read_schema.timestamp_name());
+        if self.read_schema.tsid_column().is_some() {
+            // When tsid exists, that means default primary key (tsid, timestamp) is used.
+            // So, all filters of tag columns(tsid is the hash result of all tags) can be pushed down.
+            for column in self.read_schema.columns() {
+                if column.is_tag {
+                    key_set.insert(&column.name);
+                }
+            }
+        } else {
+            // When tsid does not exist, that means user defined primary key is used.
+            // So, only filters of primary key can be pushed down.
+            for primary_idx in self.read_schema.primary_key_indexes() {
+                let primary_column = self.read_schema.column(*primary_idx);
+                key_set.insert(&primary_column.name);
+            }
+        }
```

Comment:
I guess we can encapsulate the block code into a method called `collect_unique_keys` on the type of `read_schema`.